doc/ttyreplay.1
.TH "ttyreplay" "1" "2009-09-28" "ttyrpld" "tty logging daemon suite"
.SH "Name"
.PP
ttyreplay \(em realtime log file player
.SH "Syntax"
.PP
\fBttyreplay\fP [\fB\-\-no\-pctrl\fP] [\fB\-Fft\fP] [\fB\-J\fP
\fItime\fP|\fB\-j\fP \fIpackets\fP] [\fB\-S\fP \fIfactor\fP] [\fB\-T\fP]
[\fB\-m\fP \fImsec\fP]
.SH "Description"
.PP
ttyreplay is the tool to see \(em to replay \(em what has been captured by (the
kernel module and) rpld. Give it any number of files you want to see again.
Options apply to all files.
.PP
Without any arguments, ttyreplay will try to play STDIN if it is not a tty. The
special file "-" (standalone dash) can be used to indicate STDIN, if STDIN is
not the only file to play.
.SH "Options"
.TP
\fB\-\-no\-pctrl\fP
Disable play control. When play control is enabled, you can pause replaying,
skip forwards, raise/lower speed. See below for more.
.TP
\fB\-F\fP
Live feed follow mode. Seek to the end of the file and tail-follow any new
contents that are written to it. This does the same as `tail -f` does with
normal text logfiles.
.TP
\fB\-J\fP \fItime\fP
Skim to time position before starting to play.
.TP
\fB\-S\fP \fIfactor\fP
Plays the file back with a time warp of factor. 1.0 is normal time, 2.0 is
twice as fast, 0.5 half speed, etc.
.TP
\fB\-T\fP
Shows (replays) the log without any delays. This can be used to generate a
screen log like script or screen (rpl logfile without any packet headers,
basically) for easy grepping.
.TP
\fB\-f\fP
Catch-up follow mode. Plays the file back from the start as usual and switches
into \-F mode when EOF is reached.
.TP
\fB\-j\fP \fIcount\fP
Skim \fIcount\fP packets before starting to play.
.TP
\fB\-m\fP \fImsec\fP
Set the maximum delay in milliseconds that will be executed between packets.
This check is done after the \-S option, so \-m1000 \-S2 will first crunch a
delay request (from the logfile) by factor 2, and then see if it is more than 1
second.
.TP
\fB\-t\fP
Shows the timestamp at the top right corner. Note that this might not work with
all terminals, because for one, not all support it, and second, the escape code
is currently hardcoded (no terminfo use), so it may break on some strange
terminal types.
.TP
\fB\-t \-t\fP
\fB-t\fP will use "%H:%M:%S" as the format, while \fB\-tt\fP (or \fB\-t \-t\fP)
uses "%d.%m %H:%M". The latter is useful for users on a tty which have been
logged in across a day boundary.
.SH "Player control"
.PP
While ttyreplay is running, you can use the following keys to interactively
instruct ttyreplay to do something:
.TP
(spacebar)
Toggle pause
.TP
< (less than)
Move to the previous file. Note that this might not work if
the previous one was a pipe, STDIN or similar, because such
have been emptied already.
.TP
> (greater than)
Move to the next file
.TP
y or z
Multiply playing speed by factor 0.75
.TP
x
Normal playing speed
.TP
c
Multiply playing speed by factor 1.33
.TP
[
Decrease playing speed by 0.10
.TP
]
Increase playing speed by 0.10
.TP
(keypad) 6
Skip forward 10 seconds
.TP
(keypad) 9
Skip forward 60 seconds
.TP
q or Ctrl+C
Quit
.TP
e
Toggle ECHO mode, i.e. show RPLEVT_READ packets (keypresses) as well, rather
than just screen output. This is useful for echo-less (mostly password)
prompts.
.PP
Skipping backwards is not implemented and is even impossible on pipes.
.SH "See also"
.PP
\fBrpl\fP(5), \fBrpld\fP(8)


doc/changelog.txt

Notes
=====

You do not need to repatch your kernel or recompile rpldev if they have not
changed significantly between ttyrpld releases.

│                │last time kpatch│                            │
│Operating system│(rpldhk) changed│last time rpldev changed:   │
├────────────────┼────────────────┼────────────────────────────┤
│Linux           │2.50            │2.60                        │
│FreeBSD         │2.60            │2.60                        │
│OpenBSD         │-               │2.60                        │
│NetBSD          │2.50            │2.60                        │


Changelog
=========

HEAD

Changes:
- doc: move changelog to text/plain
- rpldev: NetBSD 5.0.1 support
- rpldev: OpenBSD 4.5 support
- rpldev: Linux 2.6.31 support
- rpldev: FreeBSD 8.0 support
- packet: increase the "size" field to 32 bits
- packet: combine "magic" and "event" fields, and move to front


v2.52 (2008-09-07)

  • Updated kernel components for Linux 2.6.27
  • Updated userspace code for libHX 1.25

v2.51 (2008-05-15)

  • Updated rpldhk and rpldev for Linux 2.6.25, OpenBSD 4.3, FreeBSD 7.0,
    NetBSD 4.0

v2.50 (2008-01-01)

  • changed the packet format to be compatible between 32-bit and 64-bit
  • added rplcvt utility for converting old on-disk rpld files
  • explicitly blacklist /dev/stdout, /dev/stderr and /dev/stderr - these have
    been popping up as tty names because there is such a symlink in Linux's /
    dev. (But we need symlink processing for Solaris, eww.)

v2.19 (2007-09-17)

  • The kernel interface has been reduced (hooks removed). The 2.19 userspace
    is still compatible (throws warnings though) with 2.18 kernel components
    however. The 2.19 kernel patch is for 2.6.22.
  • an experimental Solaris kernel driver has been added - tested with Solaris
    11.snv only, but may work with older ones.
  • fix a file descriptor leak in rpld (thanks J.Eller)
  • fix a symlink recursion when walking /dev (J.Eller)
  • raise maximum number of open files (J.Eller)
  • handle permission restrictions on /dev nodes
  • fix directory name of Persian translation
  • moved to autotools
  • build with new libHX 1.10

v2.18 (2006-07-01)

  • Fixed OpenBSD 3.9 kpatch
  • Readded workaround for OpenBSD/NetBSD EINTR signal handling that got lost
    in 2.17

v2.17 (2006-06-19)

  • Support for Linux 2.6.17, FreeBSD 6.1, OpenBSD 3.9
  • added missing -fprofile-arcs -ftest-coverage to LDFLAGS for PROF=1

v2.16 (2006-04-01)

  • user/rpld.c: removed a superfluous variable
  • Added polling functions to the BSD rpldevs. This fixes the high CPU usage
    (which is due to the implementation of pthreads on OpenBSD) observed with
    rpld/OpenBSD.
  • OpenBSD part is again just a patch since modules are said to be unsupported
    on non-i386
  • Due to popular request, the docs now come with a black-on-white ("paper
    color") color scheme and no fixed font size. The old gray-on-black
    ("console color") is available as an alternate stylesheet and can be used
    in Firefox using View > Page Style > Black background.
  • Added the EVT_MAGIC pseudo event type
  • Added share/ttyrpld.magic for the file(1) utility

v2.15 (2006-03-11)

General:

  • added support for NetBSD 3.0
  • shifts for majors>255 and minors>255 were broken in include/dev.h

kpatch:

  • [All] fixed some SMP race conditions
  • [OpenBSD] fixed: rpl_deinit() was not called in ttyclose()

rpldev:

  • [All] rename functions to make function more clear (mv_to_user ->
    circular_get, mv_buffer -> circular_put_packet, mv_buffer2 -> circular_put)
  • [All] make avail_R() and avail_W() return size_t instead of ssize_t,
    addressing some potential overflow problems
  • [Linux] document circulary calculation of avail_*(), please report errors
  • [FreeBSD] merged freebsd-*/rpldev.c into one file
  • [FreeBSD/OpenBSD] removed SKIP_PTM, since master ptys do not generate any
    noise (that's a kpatch feature)
  • [OpenBSD] rpldev is now a module (rpldev_mod.o)
  • [OpenBSD] loads at new device number (228,0), please re-mknod your /dev/rpl
    node

v2.12 (2006-01-21)

  • fixed: FreeBSD 6.0 hands krpl_close() a NULL pointer in some cases; handle
    it properly
  • some doc updates
  • GNUmakefile: "all" target does not include "doc" anymore, because BSDs
    (standard installation) often ship without php, w3m and perl.
  • added the '[' and ']' keys to ttyreplay (lower/raise speed by 0.10)

v2.11 (2006-01-06)

Added support back in for the latest releases of the previous generation of
kernels. For Linux, that's 2.4.something (USE DISCOURAGED!), and for FreeBSD
that's 5.3.

v2.10 (2005-12-04)

Support for OpenBSD 3.8 and FreeBSD 6.0 was added; permission denied, syslog
spamming, cpu hog, ownership and symlink recursion bugs were fixed. Docs were
updated to reflect advances since the last seven releases.

  • Support for OpenBSD 3.8 and FreeBSD 6.0
  • Documentation has been updated to reflect advances since the last seven
    releases.
  • user/replay.c: fixed a color bug when -t was active (spotted whilst under
    FreeBSD)
  • user/*.c: use usleep(10000) instead of sched_yield() to reduce idle cpu hog
  • user/rpld.c: umask was set wrong, so 1.) logging did not take place 2.)
    syslog was spammed with permission denied warnings
  • user/rpld.c: added a rate limiter for syslog messages
  • user/rpld.c: find_devnode_dive() forgot to check for symlinks so that we
    dived into /dev/fd, which is not wanted, because /dev/fd/* itself includes
    /dev again, which lead to unwanted results.
  • user/rpld.c: fixed a filedescriptor leak in fill_info()
  • user/rpld.c: another try at getting ownerships right; rpld still had a case
    where the recorded owner differed from the owner of the /dev node (with
    respect to the design decision). See one below:
  • linux-2.6/rpldev.c: /dev/tty is a special case with respect to ownerships,
    so it must be filtered out in krpl_init() and krpl_open().

doc/ttyreplay.1.php
<?php include_once("_header.php"); ?>

<h1>ttyreplay - realtiem log file player</h1>

<p><code>ttyreplay <b>[</b>--no-pctrl<b>] [</b>-F<b>|</b>-f<b>] [</b>-J
<i>time</i><b>|</b>-j <i>packets</i><b>] [</b>-S <i>factor</i><b>] [</b>-T<b>]
[</b>-m <i>msec</i><b>] [</b>-t<b>] [</b>-?<b>|</b>--help<b>]</b></code></p>

<h1>Description</h1>

<p class="block"><i>ttyreplay</i> is the tool to see&nbsp;-- to replay&nbsp;--
what has been captured by (the kernel module and) <i>rpld</i>. Give it any
number of files you want to see again. Options apply to all files.</p>

<p class="block">Without any arguments, <i>ttyreplay</i> will try to play STDIN
if it is not a tty. The special file "<code>-</code>" (so-called <i>lone
dash</i>) can be used to indicate STDIN, if STDIN is not the only file to
play.</p>

<h1>Options</h1>

<table border="1">
  <tr>
    <td class="t1" nowrap="nowrap"><code><b>--no-pctrl</b></code></td>
    <td class="t1">Disable play control. When play control is enabled, you can
      pause replaying, skip forwards, raise/lower speed. See below for
      more.</p>
  </tr>
  <tr>
    <td class="t2" nowrap="nowrap"><code><b>-F</b></code></td>
    <td class="t2">Live feed follow mode. Seek to the end of the file and
      tail-follow any new contents that are written to it. This does the same
      as <code>tail -f</code> does with normal text logfiles.</td>
  </tr>
  <tr>
    <td class="t1" nowrap="nowrap"><code><b>-J</b> <i>time</i></code></td>
    <td class="t1">Skim to <i>time</i> position before starting to play.</td>
  </tr>
  <tr>
    <td class="t2" nowrap="nowrap"><code><b>-S</b> <i>factor</i></code></td>
    <td class="t2">Plays the file back with a time warp of <i>factor</i>.
      <code>1.0</code> is normal time, <code>2.0</code> is twice as fast,
      <code>0.5</code> half speed, etc.</td>
  </tr>
  <tr>
    <td class="t1" nowrap="nowrap"><code><b>-T</b></code></td>
    <td class="t1">Shows (replays) the log without any delays. This can be used
      to generate a screen log like <i>script</i> or <i>screen</i> (rpl logfile
      without any packet headers, basically) for easy grepping.</td>
  </tr>
  <tr>
    <td class="t2" nowrap="nowrap"><code><b>-f</b></code></td>
    <td class="t2">Catch-up follow mode. Plays the file back from the start as
      usual and switches into <code>-F</code> mode when EOF is reached.</td>
  </tr>
  <tr>
    <td class="t1" nowrap="nowrap"><code><b>-j</b> <i>count</i></code></td>
    <td class="t1">Skim <i>count</i> packets before starting to play.</td>
  </tr>
  <tr>
    <td class="t2" nowrap="nowrap"><code><b>-m</b> <i>msec</i></code></td>
    <td class="t2">Set the maximum delay in milliseconds that will be executed
      between packets. This check is done after the <code>-S</code> option, so
      <code>-m1000 -S2</code> will first crunch a delay request (from the
      logfile) by factor 2, and then see if it is more than 1 second.</td>
  </tr>
  <tr>
    <td class="t1" nowrap="nowrap"><code><b>-t</b></code></td>
    <td class="t1">Shows the timestamp at the top right corner. Note that this
      might not work with all terminals, because for one, not all support it,
      and second, the escape code is currently hardcoded (no terminfo use), so
      it may break on some strange terminal types.</p>

      <p class="block"><code>-t</code> will use "<code>%H:%M:%S</code>" as the
      format, while <code>-tt</code> (or <code>-t -t</code>) uses "<code>%d.%m
      %H:%M</code>". The latter is useful for users on a tty which have been
      logged in across a day boundary.</p>

    </td>
  </tr>
</table>

<h1>Player control</h1>

<p class="block">While <i>ttyreplay</i> is running, you can use the following
keys to interactively instruct <i>ttyreplay</i> to do something:</p>

<table>
  <tr>
    <td class="t1">(spacebar)</td>
    <td class="t1">Toggle pause</td>
  </tr>
  <tr>
    <td class="t2" nowrap="nowrap"><code>&lt;</code> (less than)</td>
    <td class="t2">Move to the previous file. Note that this might not work if
      the previous one was a pipe, STDIN or similar, because such have been
      emptied already.</td>
  </tr>
  <tr>
    <td class="t1" nowrap="nowrap"><code>&gt;</code> (greater than)</td>
    <td class="t1">Move to the next file</td>
  </tr>
  <tr>
    <td class="t2" nowrap="nowrap"><code>y</code> or <code>z</code></td>
    <td class="t2">Multiply playing speed by factor 0.75</td>
  </tr>
  <tr>
    <td class="t1"><code>x</code></td>
    <td class="t1">Normal playing speed</td>
  </tr>
  <tr>
    <td class="t2"><code>c</code></td>
    <td class="t2">Multiply playing speed by factor 1.33</td>
  </tr>
  <tr>
    <td class="t1"><code>[</code></td>
    <td class="t1">Decrease playing speed by 0.10</td>
  </tr>
  <tr>
    <td class="t2"><code>]</code></td>
    <td class="t2">Increase playing speed by 0.10</td>
  </tr>
  <tr>
    <td class="t1">(keypad) <code>6</code></td>
    <td class="t1">Skip forward 10 seconds</td>
  </tr>
  <tr>
    <td class="t2">(keypad) <code>9</code></td>
    <td class="t2">Skip forward 60 seconds</td>
  </tr>
  <tr>
    <td class="t1" nowrap="nowrap"><code>q</code> or <code>Ctrl+C</code></td>
    <td class="t1">Quit</td>
  </tr>
  <tr>
    <td class="t2"><code>e</code></td>
    <td class="t2">Toggle ECHO mode, i.e. show <code>EVT_READ</code> packets
      (keypresses) as well, rather than just screen output. This is useful for
      echo-less (mostly password) prompts.</td>
  </tr>
</table>

<p class="block">Skipping backwards is not implemented and is even impossible
on pipes.</p>

<h1>See also</h1>

<ul>
  <li><a href="rpl.5.php"><code>rpl(5)</code></li>
  <li><a href="rpld.8.php"><code>rpld(8)</code></li>
</ul>

<?php include_once("_footer.php"); ?>
